mountpoint-s3/src/inode/readdir.rs

```diff
@@ -0,0 +1,454 @@
+//! Utilities for implementing the `readdir` operation on a directory inode.
+//!
+//! `readdir` is conceptually simple ("just call ListObjectsV2") but has some subtleties that make
+//! it complicated:
+//!
+//! 1. It's possible for a directory to contain both a subdirectory and a file of the same name, in
+//!    which case we need to implement shadowing in the right way.
+//! 2. A directory can also contain local files/subdirectories, which we want to have appear in the
```

Comment:
Shadowing means when there's a file and directory of the same name, we need to make a choice about which one to show, because file systems aren't allowed to do that. We choose to show directories over files if both have the same name.